packages/shared/src/events/eventing.ts

```diff
@@ -19,40 +19,62 @@ export type ErrorEvent = CommonEventProps;
 export type StaleEvent = CommonEventProps;
 export type ConfigChangeEvent = CommonEventProps & { flagsChanged?: string[] };
 
-type EventMap = {
-  [AllProviderEvents.Ready]: ReadyEvent;
-  [AllProviderEvents.Error]: ErrorEvent;
-  [AllProviderEvents.Stale]: StaleEvent;
-  [AllProviderEvents.ContextChanged]: CommonEventProps;
-  [AllProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
+type ServerEventMap = {
+  [ServerProviderEvents.Ready]: ReadyEvent;
+  [ServerProviderEvents.Error]: ErrorEvent;
+  [ServerProviderEvents.Stale]: StaleEvent;
+  [ServerProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
 };
 
-export type EventContext<  U extends Record<string, unknown> = Record<string, unknown>
-> = EventMap[AllProviderEvents] & U;
+type ClientEventMap = {
+  [ClientProviderEvents.Ready]: ReadyEvent;
+  [ClientProviderEvents.Error]: ErrorEvent;
+  [ClientProviderEvents.Stale]: StaleEvent;
+  [ClientProviderEvents.ConfigurationChanged]: ConfigChangeEvent;
+  [ClientProviderEvents.ContextChanged]: CommonEventProps;
+};
+
+export type NotServerChangeEvents = ServerProviderEvents.Ready | ServerProviderEvents.Error | ServerProviderEvents.Stale;
+export type NotClientChangeEvents = ClientProviderEvents.Ready | ClientProviderEvents.Error | ClientProviderEvents.Stale | ClientProviderEvents.ContextChanged;
+export type NotChangeEvents = NotServerChangeEvents | NotClientChangeEvents;
+
+export type EventContext<
+  U extends Record<string, unknown> = Record<string, unknown>,
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = (T extends ClientProviderEvents ? ClientEventMap[T] : T extends ServerProviderEvents ? ServerEventMap[T] : never) &
+  U;
 
-export type EventDetails = EventContext & CommonEventDetails;
-export type EventHandler = (eventDetails?: EventDetails) => Promise<unknown> | unknown;
+export type EventDetails<
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = EventContext<Record<string, unknown>, T> & CommonEventDetails;
+export type EventHandler<
+  T extends ServerProviderEvents | ClientProviderEvents = ServerProviderEvents | ClientProviderEvents,
+> = (eventDetails?: EventDetails<T>) => Promise<unknown> | unknown;
 
-export interface Eventing {
+export interface Eventing<T extends ServerProviderEvents | ClientProviderEvents> {
```

Comment:
Made this generic to support either client OR server events in the implementation.